File: C++/Arrays/ArrayList/ArrayList.cpp
/*******************************************************
 *  @file ArrayList.cpp
 *  @author Stephen Hall
 *  @date 11/22/17.
 *  @copyright 2017 Stephen Hall. All rights reserved.
 *  @details ArrayList implementation in C++
 ********************************************************/

/* Only needed because I did not include a file where it is defined */
#ifndef NULL
#define NULL   ((void *) 0)
#endif

#include "ArrayList.h"

/**
 * ArrayList default constructor
 * @tparam T: Generic type
 */
template <typename T>
ArrayList::ArrayList()
{
    array = new T[size = 4];
    for(int i = 0; i < size; i++)
        array[i] = NULL;
    count = 0;
}

/**
 * ArrayList Constructor initialized to given size
 * @tparam T: Generic Type
 * @param size: Size to initialize the array list to
 */
template <typename T>
ArrayList::ArrayList(unsigned int size)
{
    array =  new T[this->size = size > 0 ? size : 4];
    for(int i = 0; i < size; i++)
        array[i] = NULL;
    count = 0;
}

/**
 * Adds data into the ArrayList
 * @tparam T: Generic Type
 * @param data: data to add into the ArrayList
 * @return T: Data sdded into the ArrayList
 */
template <typename T>
T ArrayList::add(T data)
{
    if(data == NULL)
        return NULL;

    if(count == size)
        resize();

    array[count] = data;
    count++;

    return data;
}

/**
 * Appends an Array of T onto the ArrayList
 * @tparam T: Generic Type
 * @param dataArray: array to append
 * @param size: size of the array
 * @return bool: success|fail
 */
template <typename T>
bool ArrayList::append(T* dataArray, int size)
{
    if(dataArray != NULL){
        for(int i = 0; i < size; i++)
        {
            if (dataArray[i] != NULL)
                add(dataArray[i]);
        }
        return true;
    }
    return false;
}

/**
 * Clears all the elements in the Array List but keeps the size intact
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayList::clear()
{
    for(int i = 0; i < size; i++)
        array[i] = NULL;
}

/**
 * ArrayList Destructor
 * @tparam T: Generic Type
 */
template <typename T>
ArrayList::~ArrayList()
{
    delete [] array;
    array = NULL;
}

/**
 * Gets the number of elements in the ArrayList
 * @return
 */
int ArrayList::length()
{
    return count;
}

/**
 * Overloaded [] operator for array indexing
 * @tparam T: Generic type
 * @param index: index to access
 * @return T: data at the given index or null
 */
template <typename T>
T& ArrayList::operator[](int index)
{
    return (index < size) ? array[index] : NULL;
}

/**
 * Removes the first instance of the given data from the ArrayList
 * @tparam T: generic type
 * @param data: Data to remove
 * @return T: data removed from the ArrayList
 */
template <typename T>
T ArrayList::remove(T data)
{
    for(int i = 0; i < count; i++)
    {
        if(array[i] == data)
        {
            T tmp = array[i];
            array[i] = NULL;
            return tmp;
        }
    }
    return NULL;
}

/**
 * Resets the array to its default state
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayList::reset()
{
    delete [] array;
    array = new T[size = 4];
    for(int i = 0; i < size; i++)
        array[i] = NULL;
    count = 0;
}

/**
 * Resize's the ArrayLists internal array
 * @tparam T: Generic Type
 */
template <typename T>
void ArrayList::resize()
{
    size *= 2;
    T* tmp =  new T[size];
    for(int i = 0; i < length(); i++)
        tmp[i] = array[i];

    delete [] array;
    array = tmp;
    array = tmp;
}

File: PHP/Arrays/ArrayList/ArrayList.php
<?php
/*******************************************************
 *  ArrayList.java
 *  Created by Stephen Hall on 11/06/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  ArrayList implementation in Java
 ********************************************************/

/**
 * ArrayList Class
 */
class ArrayList {

    /**
     * Private Members
     */
    private $array = [];
    private $count = 0;
    private $size = 0;


    /**
     * ArrayList constructor initialized to a specific size
     * @param $size : Size to initialize the array to
     */
     function __construct($size){
        $this->count = 0;
        if($size > 0)
            $this->size = $size;
        else
            $this->size = 4;
        $this->array = [$this->size];
    }

    /**
     * Doubles the size of the internal array
     */
    private function resize(){
        $tmp = [$this->size * 2];
        for($i = 0; $i < $this->count; $i++){
            $tmp[$i] = $this->array[$i];
         }
        $this->size *= 2;
        $this->array = $tmp;
    }

    /**
     * Adds new item into the array
     * @param $data : Data to add into the array
     * @return mixed : data added into the array
     */
    public function add($data){
        if(!isset($data))
            return null;

        if($this->count == $this->size)
            $this->resize();

        $this->array[$this->count] = $data;
        $this->count++;

        return $data;
    }

    /**
     * Appends the contents of an array to the ArrayList
     * @param $data : array to append
     * @return boolean : success|fail
     */
    public function append($data){
        if(isset($data)){
            foreach($data as $aData) {
                if (isset($aData))
                    $this->add($aData);
            }
            return true;
        }
        return false;
    }

    /**
     * Sets the data at the given index
     * @param $index : index to set
     * @param $data : data to set index to
     * @return boolean: success|fail
     */
    public function set($index, $data){
        if($index >= 0 && $index < $this->size) {
            $this->array[$index] = $data;
            return true;
        }
        return false;
    }

    /**
     * Gets the data at the arrays given index
     * @param $index : Index to get data at
     * @return mixed : Data at the given index or default value of T if index does not exist
     */
    public function get($index){
        return ($index >= 0 && $index < $this->size) ? $this->array[$index] : null;
    }

    /**
     * Removes the data at arrays given index
     * @param $index : Index to remove
     * @return mixed : Data removed from the array or null if index does not exist
     */
    public function remove($index){
        if ($index < 0 || $index > $this->count)
            return null;

        $tmp = $this->array[$index];
        $this->array[$index] = null;
        $this->count--;
        return $tmp;
    }

    /**
     * Resets the internal array to default size with no data
     */
    public function reset(){
        $this->count = 0;
        $this->size = 4;
        $this->array = [$this->size];
    }

    /**
     * Clears all data in the array leaving size intact
     */
    public function clear(){
        for($i = 0; $i < $this->count; $i++){
            $this->array[$i] = null;
    }
        $this->count = 0;
    }

    /**
     * Gets the current count of the array
     * @return int : Number of items in the array
     */
    public function count(){
        return $this->count;
    }
}
